stuff/canonize_tags.py

```diff
@@ -0,0 +1,138 @@
+
+import MySQLdb.constants.ER
+
+import cron.lib.db as db
+import cron.lib.logs as logs
+
+LOCALIZATIONS_ROOT = os.path.abspath(os.path.join(__file__,
+                                                  '..', '..',
+                                                  'frontend/www/js/omegaup'))
+LANGS = ['en', 'es', 'pt']
+
+
+def normalize_tag(tag: str) -> str:
+    '''Normalizes tags, similar to Tags::normalize() in PHP'''
+    # Remove empty spaces
+    tag = tag.strip()
+    # Remove accents
+    tag = unicodedata.normalize('NFD',
+                                tag).encode('ascii', 'ignore').decode("utf-8")
+    # Use '-' for splitting if necessary
+    tag = re.sub(r'[^a-z0-9]', '-', tag.lower())
+    tag = re.sub(r'--+', '-', tag)
+    return tag
+
+
+def insert_new_tags(tags: Set[str],
+                    cur: MySQLdb.cursors.DictCursor) -> None:
+    '''Inserts new problem tags inside Tags table on DB'''
+    logging.info('Inserting new Tags on database')
+    cur.executemany('''INSERT IGNORE INTO `Tags`(`name`)
+                    VALUES (%s);''',
+                    [(tag,) for tag in tags])
+
+
+def get_inverse_mapper(
+        cur: MySQLdb.cursors.DictCursor) -> Mapping[str, Mapping[str, str]]:
+    '''Gets the inverse mapper for problem tags entries in lang files'''
+    logging.info('Getting tags mapper from lang files.')
+    new_tags = set()
+    inverse_mapper = {}
+    for lang in LANGS:
+        path = os.path.join(LOCALIZATIONS_ROOT, 'lang.%s.json' % (lang))
+        with open(path, 'r') as f:
+            mappings = json.load(f)
+            inverse_mapping = {}
+            for key, value in mappings.items():
+                if key.startswith('problemTopic'):
+                    normalized_tag = normalize_tag(key)
+                    new_tags.add(normalized_tag)
+                    inverse_mapping[normalize_tag(value)] = normalized_tag
+            inverse_mapper[lang] = inverse_mapping
+    insert_new_tags(new_tags, cur)
+    return inverse_mapper
+
+
+def migrate_tags(cur: MySQLdb.cursors.DictCursor,
+                 mapper: Mapping[str, Mapping[str, str]]) -> None:
+    '''Reads all suggestions and modifies their tags if necessary'''
+    cur.execute('''SELECT `qualitynomination_id`, `contents`
+                    FROM `QualityNominations`
+                    WHERE `nomination` = 'suggestion';''')
+    to_update = []
+    for row in cur:
+        try:
+            contents = json.loads(row['contents'])
+        except json.JSONDecodeError:  # pylint: disable=no-member
+            logging.exception('Failed to parse contents')
+            continue
+        if not contents.get('tags'):
+            continue
+        nomination_id = row['qualitynomination_id']
+        canonized_tags = []
+        for tag in contents['tags']:
+            canonized_tags.append(tag)
+            for lang in LANGS:
+                if mapper[lang].get(tag):
+                    canonized_tags[-1] = mapper[lang][tag]
```

Comment:
aquí puedes mejor hacer

```suggestion
            for lang in LANGS:
                if mapper[lang].get(tag):
                    canonicalized_tags.append(mapper[lang][tag])
                    break
            else:  # sí, este else está en el for, no en el if.
                canonicalized_tags.append(tag)
```

también, bajo qué circunstancia se ejecutaría el else? no era el objetivo de esto el únicamente permitir tags i18n-ados?

Reply:
Tienes razón. No debería ejecutar ese else.